src/Microsoft.AspNetCore.Blazor.Server/Circuits/CircuitSynchronizationContext.cs

```diff
@@ -0,0 +1,269 @@
+            // the call path of ThreadPool.QueueUserWorkItem and System.Threading.QueueUserWorkItemCallback.
+            ExecutionContext executionContext = null;
+            if (!ExecutionContext.IsFlowSuppressed())
+            {
+                executionContext = ExecutionContext.Capture();
+            }
+
+            return antecedant.ContinueWith(BackgroundWorkThunk, new WorkItem()
+            {
+                SynchronizationContext = this,
+                ExecutionContext = executionContext,
+                Callback = d,
+                State = state,
+            }, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Current);
+        }
+
+        private void ExecuteSynchronously(
+            TaskCompletionSource<object> completion,
+            SendOrPostCallback d,
+            object state)
+        {
+            var original = Current;
+            try
+            {
+                SetSynchronizationContext(this);
+                _state.IsBusy = true;
+
+                d(state);
+            }
+            finally
+            {
+                _state.IsBusy = false;
+                SetSynchronizationContext(original);
+
+                completion?.SetResult(null);
+            }
+        }
+
+        private void ExecuteBackground(WorkItem item)
+        {
+            if (item.ExecutionContext == null)
+            {
+                try
+                {
+                    ExecuteSynchronously(null, item.Callback, item.State);
+                }
+                catch (Exception ex)
+                {
+                    DispatchException(ex);
+                }
+
+                return;
+            }
+
+            // Perf - using a static thunk here to avoid a delegate allocation.
+            try
+            {
+                ExecutionContext.Run(item.ExecutionContext, ExecutionContextThunk, item);
+            }
+            catch (Exception ex)
+            {
+                DispatchException(ex);
+            }
+        }
+
+        private void DispatchException(Exception ex)
+        {
+            var handler = UnhandledException;
+            if (handler != null)
+            {
+                handler(this, new UnhandledExceptionEventArgs(ex, isTerminating: false));
+            }
+        }
+
+        private class State
+        {
+            public bool IsBusy; // Just for debugging
+            public object Lock = new object();
+            public Task Task = Task.CompletedTask;
+
+            public override string ToString()
+            {
+                return $"{{ Busy: {IsBusy}, Pending Task: {Task} }}";
+            }
+        }
+
+        private class WorkItem
+        {
+            public CircuitSynchronizationContext SynchronizationContext;
+            public ExecutionContext ExecutionContext;
+            public SendOrPostCallback Callback;
+            public object State;
```

Comment:
I don't totally follow why this `State` property exists, as it seems like that only gets passed to `ExecuteSynchronously`, which can already get the same data from `_state`. Of course I might be wrong as it's hard to track execution from reading the PR.

Reply:
There are cases where the runtime makes a copy of a sync context. We want all of those copies to share state. 

I'm not actually sure of the details of how and when this occurs - but this is a pattern that isolates us from it.